pkg/controller/rsm2/replica_util.go

```diff
@@ -439,6 +465,82 @@ func buildReplicaTemplateGroups(rsm *workloads.ReplicatedStateMachine) map[strin
 	return replicaTemplateGroups
 }
 
+func getInstanceTemplateMap(annotations map[string]string) (map[string]string, error) {
+	if annotations == nil {
+		return nil, nil
+	}
+	templateRef, ok := annotations[templateRefAnnotationKey]
+	if !ok {
+		return nil, nil
+	}
+	templateMap := make(map[string]string)
+	if err := json.Unmarshal([]byte(templateRef), &templateMap); err != nil {
+		return nil, err
+	}
+	return templateMap, nil
+}
+
+func getInstanceTemplates(rsm *workloads.ReplicatedStateMachine, tree *kubebuilderx.ObjectTree) []workloads.InstanceTemplate {
+	if tree == nil {
+		return rsm.Spec.Instances
+	}
+	templateMap, err := getInstanceTemplateMap(rsm.Annotations)
+	// error has been checked in prepare stage, there should be no error occurs
+	if err != nil {
+		return nil
+	}
+	findTemplate := func() (*corev1.ConfigMap, error) {
+		for name, templateName := range templateMap {
+			if name != rsm.Name {
+				continue
+			}
+			// find the compressed instance templates, parse them
+			template := builder.NewConfigMapBuilder(rsm.Namespace, templateName).GetObject()
+			templateObj, err := tree.Get(template)
+			if err != nil {
+				return nil, err
+			}
+			template, _ = templateObj.(*corev1.ConfigMap)
+			return template, nil
+		}
+		return nil, nil
+	}
+
+	template, err := findTemplate()
+	// error should not occur as has been checked in tree load stage.
+	if err != nil {
+		return nil
+	}
+	if template == nil {
+		return rsm.Spec.Instances
+	}
+
+	// if template is found with incorrect format, try it as the whole templates is corrupted.
+	if template.BinaryData == nil {
+		return nil
+	}
+	templateData, ok := template.BinaryData[templateRefDataKey]
+	if !ok {
+		return nil
+	}
+	reader, err := zstd.NewReader(nil)
```

Comment:
You can use a global reader or writer for avoiding frequent GC with large objects.

Reply:
good point.

Reply:
fixed in: [a049da7](https://github.com/apecloud/kubeblocks/pull/6865/commits/a049da7087d7e40577eaed1b1892c60b69782484)